### PR TITLE
Add Gumbel-Softmax sampling function

### PR DESCRIPTION
**Pull request by ishihara1989**
This PR adds function which draws samples from Gumbel-Softmax distribution, described in [Categorical Reparameterization with Gumbel-Softmax](https://arxiv.org/abs/1611.01144).

### CODE REVIEW

File: tests/chainer_tests/functions_tests/noise_tests/test_gumbel_softmax.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import chainer
+from chainer import cuda
+from chainer import functions
+from chainer import testing
+from chainer.testing import attr
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 2), ],
+}))
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/noise_tests/test_gumbel_softmax.py]: Can you add other dtypes? `'dtype': [numpy.float16, numpy.float32, numpy.float64]`

File: tests/chainer_tests/functions_tests/noise_tests/test_gumbel_softmax.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import chainer
+from chainer import cuda
+from chainer import functions
+from chainer import testing
+from chainer.testing import attr
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 2), ],
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/noise_tests/test_gumbel_softmax.py]: Add `()`

File: chainer/functions/noise/gumbel_softmax.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from chainer import cuda
+from chainer.functions.activation.softmax import softmax
+
+
+def gumbel_softmax(log_pi, tau=0.1, axis=1):
```
Comment (unnonouno) [inline chainer/functions/noise/gumbel_softmax.py]: Why not using `Variable` for `tau`?

File: chainer/functions/noise/gumbel_softmax.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+from chainer import cuda
+from chainer.functions.activation.softmax import softmax
```
Comment (unnonouno) [inline chainer/functions/noise/gumbel_softmax.py]: Don't import function directly. Instead import module.
```
from chainer.functions.activation import softmax
...
softmax.softmax
```

File: chainer/functions/noise/gumbel_softmax.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def gumbel_softmax(log_pi, tau=0.1, axis=1):
+    """Gumbel-Softmax sampling function.
+
+    This function draws samples :math:`y_i` from Gumbel-Softmax distribution,
+        :math:`y_i = {\\exp((g_i + \\log\\pi_i)/\\tau) \
```
Comment (unnonouno) [inline chainer/functions/noise/gumbel_softmax.py]: How about using  `.. math::`?

File: chainer/functions/noise/gumbel_softmax.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                    \\over \\sum_{j}\\exp((g_j + \\log\\pi_j)/\\tau)}`,
+        where :math:`\\tau` is a temperature parameter and \
```
Comment (unnonouno) [inline chainer/functions/noise/gumbel_softmax.py]: Backslashes at end of lines are not required.

File: chainer/functions/noise/gumbel_softmax.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            :math:`g_i` s are samples drawn from \
+            Gumbel distribution :math:`Gumbel(0, 1)`
+
+    See `Categorical Reparameterization with Gumbel-Softmax \
+    <https://arxiv.org/abs/1611.01144>`_.
+
+    Args:
+        log_pi (:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
```
Comment (unnonouno) [inline chainer/functions/noise/gumbel_softmax.py]: <pre>:class:</pre> is not required. Types are automatically parsed here.

File: chainer/functions/noise/gumbel_softmax.py (hunk 3 above)
Comment (ishihara1989) [reply chainer/functions/noise/gumbel_softmax.py]: Maybe I'm misunderstanding but I couldn't get your point.
In my understanding, when `Variable` is passed, it can be properly backproped because current implementation is just combination of arithmetic and `Function` call.
If backprop is not needed scalar is enough.

**Comment by ishihara1989** [comment]
Sorry for replying late. I'm beginner on python document generation system and was struggling against it :(

**Comment by unnonouno** [comment]
jenkins test this please

File: chainer/functions/noise/gumbel_softmax.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+from chainer import cuda
+from chainer.functions.activation import softmax
```
Comment (beam2d) [inline chainer/functions/noise/gumbel_softmax.py]: Use `chainer.functions.softmax` instead of `softmax.softmax` (it is allowed inside of any functions).

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
LGTM
